Add 'db' package that handles operations on SQLite3

jzigmund opened the pull request:
Create Database "API" for storing and receiving "data" that is passed to and produced by actors

Commit:
Add 'db' package that handles operations on SQLite3

mfranczy commented on pkg/db/db.go:
```diff
@@ -0,0 +1,138 @@
+		return nil, err
+	}
+	db := (*Database)(conn)
+	if err := db.createTable(); err != nil {
```
I think this is not the best idea, if you want to do something like this, it's better to check on start.

mfranczy commented on pkg/db/db.go:
```diff
@@ -0,0 +1,138 @@
+
+func (db *Database) createTable() error {
+	sql := `
+	  CREATE TABLE IF NOT EXISTS migrations(
```
Take this as my personal opinion and suggestion.
I don't like to execute DDL inside app, I understand that structure is small and it's not a big deal, although it's bad to keep DDL and DML in one.

jzigmund commented:
updated PR, fixed golint issues, added ```func init()```

jzigmund commented on pkg/db/db.go:
```diff
@@ -0,0 +1,149 @@
+	_ "github.com/mattn/go-sqlite3"
+)
+
+const dbName string = "leapp.db"
```
/var/lib/leapp/leapp.db

jzigmund commented on pkg/db/db.go:
```diff
@@ -0,0 +1,149 @@
+func init() {
+	conn, err := Connect(dbName)
+	if err != nil {
+		fmt.Println(err)
```
panic here

jzigmund commented on pkg/db/db.go:
```diff
@@ -0,0 +1,149 @@
+	if err != nil {
+		fmt.Println(err)
+	}
+
```
defer conn.Close()

jzigmund commented on pkg/db/db.go:
```diff
@@ -0,0 +1,149 @@
+	}
+
+	if err := conn.createTable(); err != nil {
+		conn.Close()
```
remove

jzigmund commented on pkg/db/db.go:
```diff
@@ -0,0 +1,149 @@
+
+	if err := conn.createTable(); err != nil {
+		conn.Close()
+		fmt.Println(err)
```
panic here

jzigmund commented on pkg/db/db.go:
```diff
@@ -0,0 +1,149 @@
+}
+
+//Connect connects to DB with given name database name
+func Connect(dbName string) (*Database, error) {
```
```golang
func connect(dbName string) (*Database, error)
```
 and big one without param
```golang
func Connect() (*Database, error) {
    return connect(dbName)
}
```

jzigmund commented on pkg/db/db.go:
```diff
@@ -0,0 +1,149 @@
+func (db *Database) Close() error {
+	return db.sql().Close()
+}
+
```
comment temporary until packaging will solve this

jzigmund commented on pkg/db/db.go:
```diff
@@ -0,0 +1,149 @@
+	UID, Name, Host, dataType, Data string
+}
+
+func init() {
```
comment temporary until packaging will solve this

jzigmund commented on pkg/db/db_test.go:
```diff
@@ -0,0 +1,92 @@
+}
+
+func setup() {
+	testConn, _ = Connect(testDbName)
```
use connect (small) and ":memory:" (drop testDbName)

jzigmund commented:
@vinzenz @mfranczy updated PR, add commit that introduces ```db``` package into ```leapp-deamon``` and necessary stuff to fake (skip) an assurance that DB directory exist, DB connection for test purpose.

vinzenz commented on pkg/db/db.go:
```diff
@@ -0,0 +1,155 @@
+
+//EnsureDbPath checks wether the leapp directory exists, if not it tries to create it
+//temporary function, will be solved in packaging later
+func EnsureDbPath() {
```
this should not panic but return the error here - And if it really is temporary that should be with the caller directly

vinzenz commented on pkg/db/db.go:
```diff
@@ -0,0 +1,155 @@
+
+//Connect connects to DB
+func Connect() *Database {
+	conn := connect(dbDir + dbName)
```
Please do use filepath.Join(dbDir, dbName) from the "path/filepath" package - This code can introduce bugs otherwise if someone changes for example the dbDir contant and forgets to add the pending slash

vinzenz commented on pkg/db/db.go:
```diff
@@ -0,0 +1,155 @@
+	return db.sql().Close()
+}
+
+//CreateTable is temporary function and will be solved in packaging later
```
Since this is temporary move it to the Main where all the other temporaries are supposed to be

vinzenz commented on pkg/db/db_test.go:
```diff
@@ -0,0 +1,91 @@
+
+func setup() {
+	testConn = connect(testDbPath)
+	testConn.CreateTable()
```
Instead of using CreateTable just execute it in the test directly and I don't think that the testDbPath constant is necessary in that case, or is it?

bertinatto commented on cmd/leapp-daemon/main.go:
```diff
@@ -18,12 +19,38 @@ var (
+//EnsureConnectCreateDb ensures that dbDir is created, connects to DB and create table
+//temporary function
+func EnsureConnectCreateDb() *db.Database {
+	if err := os.MkdirAll("/var/lib/leapp", os.ModePerm); err != nil {
```
`os.ModePerm` is actually `0777`: https://golang.org/src/os/types.go?s=1131:1151#L25

I would recommend using something less permissive, e.g., `0755`.

bertinatto commented on pkg/db/db.go:
```diff
@@ -0,0 +1,126 @@
+)
+
+//Database type is sql.DB
+type Database sql.DB
```
IMHO this isn't the best option in this case. For example, `sql()` method has to cast this type every time that it's called.

I think the best option would be to create a struct with a `sql.DB` field.

vinzenz replied:
Not sure why the cast matters to you so much - It was my suggestion to use it like this because I don't see the need for a complete new object if the same works as well

bertinatto commented on pkg/db/db.go:
```diff
@@ -0,0 +1,126 @@
+
+//Insert executes SQL INSERT query with a given record
+func (db *Database) Insert(rec []interface{}) error {
+	tx, err := db.sql().Begin()
```
It seems like this `err` variable isn't being checked.

bertinatto reviewed (commented):
I put together some suggestions in a separate branch here:

https://github.com/leapp-to/leapp-go/compare/master...bertinatto:db?expand=1

jzigmund commented:
@bertinatto updated PR, incorporated your suggestions as we spoke about them

bertinatto commented on pkg/db/db_test.go:
```diff
@@ -0,0 +1,109 @@
+func setup() {
+	conn, err := sql.Open("sqlite3", ":memory:")
+	testDB = (*Database)(conn)
+	testDBErr = err
```
This might be a nitpick, but I'll mention it either way. 😄 

I understand why you are not checking `err` here. However, if `err  != nil`, line 36 will fail and the test will exit with a message`failed to create table`, which is misleading.

IMO we should "fail fast" here.

bertinatto commented on pkg/db/db_test.go:
```diff
@@ -0,0 +1,109 @@
+
+func teardown() {
+	if err := testDB.Close(); err != nil {
+		fmt.Fprintf(os.Stderr, "failed to close DB connection: %v", err)
```
Missing `\n` :)

bertinatto commented on pkg/db/db_test.go:
```diff
@@ -0,0 +1,109 @@
+		t.Errorf("Connection is nil")
+	}
+	if testDBErr != nil {
+		t.Errorf("Connection to the DB failed: %v", testDBErr)
```
s/Errorf/Fatalf/g

`Fatalf` won't allow the execution of other tests. Since there is no a valid connection, it makes sense to fail the test right here.

bertinatto commented on pkg/db/db_test.go:
```diff
@@ -0,0 +1,109 @@
+	err := testDB.Insert(rec)
+	if err != nil {
+		t.Errorf("Insert failed: %v", err)
+	}
```
IMO it'd be nice to perform a select here and check if the data is in fact store in the DB.

But I'm OK with leaving it for another iteration.

shaded-enmity commented:
I like @bertinatto suggestion about the struct -- without all the casts the code is much easier to read and understand.

jzigmund commented:
updated PR, removed conversion of ```*Database``` to ```*sql.DB``` type and replaced that with Database struct as was proposed by @bertinatto (and supported by other developers)
